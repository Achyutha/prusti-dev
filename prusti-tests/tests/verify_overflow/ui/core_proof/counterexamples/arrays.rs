// compile-flags: -Punsafe_core_proof=true -Pcounterexample=true

use prusti_contracts::*;

//a counterexample can only be generated for directly accessed elements of a sequence

fn test1() {
    let mut a = [1; 2];
    a[0] = 2;
    assert!(a[1] == 2);
}

fn test2() {
    let a = [1; 10]; 
    //This is an exception. Here a counterexample contains all elements of the array 
    //as long as no elements is directly accessed. This works only with this syntax.
    assert!(false)
}

#[ensures(result)]
fn test3() -> bool {
    let a = [1, 2, 3, 4];
    a[0] == a[1]   
}

fn main() {}
