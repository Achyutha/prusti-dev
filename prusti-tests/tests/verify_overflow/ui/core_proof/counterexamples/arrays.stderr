error: [Prusti: verification error] the asserted expression might not hold
  --> $DIR/arrays.rs:10:5
   |
10 |     assert!(a[1] == 2);
   |     ^^^^^^^^^^^^^^^^^^
   |
note: the failing assertion is here
  --> $DIR/arrays.rs:10:5
   |
10 |     assert!(a[1] == 2);
   |     ^^^^^^^^^^^^^^^^^^
note: counterexample for "a"
       value:   [?, 1, ]
  --> $DIR/arrays.rs:8:17
   |
8  |     let mut a = [1; 2];
   |                 ^^^^^^
note: counterexample for "a"
       value:   [2, 1, ]
  --> $DIR/arrays.rs:9:5
   |
9  |     a[0] = 2;
   |     ^^^^^^^^
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

error: [Prusti: verification error] the asserted expression might not hold
  --> $DIR/arrays.rs:17:5
   |
17 |     assert!(false)
   |     ^^^^^^^^^^^^^^
   |
note: the failing assertion is here
  --> $DIR/arrays.rs:17:5
   |
17 |     assert!(false)
   |     ^^^^^^^^^^^^^^
note: counterexample for "a"
       value:   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ]
  --> $DIR/arrays.rs:14:13
   |
14 |     let a = [1; 10]; 
   |             ^^^^^^^
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/arrays.rs:21:15
   |
21 | fn test3() -> bool {
   |               ^^^^
   |
note: the error originates here
  --> $DIR/arrays.rs:21:1
   |
21 | / fn test3() -> bool {
22 | |     let a = [1, 2, 3, 4];
23 | |     a[0] == a[1]   
24 | | }
   | |_^
note: counterexample for "a"
       value:   [1, 2, ?, ?, ]
  --> $DIR/arrays.rs:22:13
   |
22 |     let a = [1, 2, 3, 4];
   |             ^^^^^^^^^^^^
note: counterexample for "result"
       value:   false
  --> $DIR/arrays.rs:23:5
   |
23 |     a[0] == a[1]   
   |     ^^^^^^^^^^^^

error: aborting due to 3 previous errors

